pkg/operator/ceph/cluster/predicate.go

```diff
@@ -21,19 +21,37 @@ import (
 	"context"
 
 	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
 	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
 	"github.com/rook/rook/pkg/clusterd"
 	discoverDaemon "github.com/rook/rook/pkg/daemon/discover"
 	"github.com/rook/rook/pkg/operator/ceph/controller"
 	"github.com/rook/rook/pkg/operator/k8sutil"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
+func compareNodes(objOld, objNew *corev1.Node) string {
+	// do not watch node if only resourceversion got changed
+	resourceQtyComparer := cmpopts.IgnoreFields(v1.ObjectMeta{}, "ResourceVersion")
+	diff := cmp.Diff(objOld.Spec, objNew.Spec, resourceQtyComparer)
+
+	// do not watch node if only LastHeartbeatTime got changed
+	resourceQtyComparer = cmpopts.IgnoreFields(corev1.NodeCondition{}, "LastHeartbeatTime")
+	diff1 := cmp.Diff(objOld.Spec, objNew.Spec, resourceQtyComparer)
+
+	if diff == "" && diff1 == "" {
+		return diff
+	}
```

Comment:
I'm not certain about this logic. It sets up 2 different comparers separately, instead of using the comparers together to ignore both at once. Even if this is working as intended, I'd like to see a unit test for the function that proves it's working as it should be.

Reply:
It seems working, wil also add the unit test
Testing:

```
2024-05-10 09:47:09.009607 D | ceph-cluster-controller: only changes in the ResourceVersion or LastHeartbeatTime "  &v1.Node{\n  \tTypeMeta: {},\n  \tObjectMeta: v1.ObjectMeta{\n  \t\t... // 3 identical fields\n  \t\tSelfLink:          \"\",\n  \t\tUID:               \"245242a2-381e-4f4c-ae5d-5d055f2e7f34\",\n- \t\tResourceVersion:   \"75120\",\n+ \t\tResourceVersion:   \"75475\",\n  \t\tGeneration:        0,\n  \t\tCreationTimestamp: {Time: s\"2024-05-06 11:53:46 +0000 UTC\"},\n  \t\t... // 4 identical fields\n  \t\tOwnerReferences: nil,\n  \t\tFinalizers:      nil,\n  \t\tManagedFields: []v1.ManagedFieldsEntry{\n  \t\t\t... // 3 identical elements\n  \t\t\t{Manager: \"kube-controller-manager\", Operation: \"Update\", APIVersion: \"v1\", Time: s\"2024-05-08 11:25:59 +0000 UTC\", ...},\n  \t\t\t{Manager: \"kube-controller-manager\", Operation: \"Update\", APIVersion: \"v1\", Time: s\"2024-05-08 11:26:00 +0000 UTC\", ...},\n  \t\t\t{\n  \t\t\t\tManager:     \"kubelet\",\n  \t\t\t\tOperation:   \"Update\",\n  \t\t\t\tAPIVersion:  \"v1\",\n- \t\t\t\tTime:        s\"2024-05-10 09:42:02 +0000 UTC\",\n+ \t\t\t\tTime:        s\"2024-05-10 09:47:08 +0000 UTC\",\n  \t\t\t\tFieldsType:  \"FieldsV1\",\n  \t\t\t\tFieldsV1:    &{Raw: `{\"f:metadata\":{\"f:annotations\":{\"f:csi.volume.kubernetes.io/node`...},\n  \t\t\t\tSubresource: \"status\",\n  \t\t\t},\n  \t\t},\n  \t},\n  \tSpec: {PodCIDR: \"10.244.0.0/24\", PodCIDRs: {\"10.244.0.0/24\"}},\n  \tStatus: v1.NodeStatus{\n  \t\tCapacity:    {s\"cpu\": {i: {...}, s: \"2\", Format: \"DecimalSI\"}, s\"ephemeral-storage\": {i: {...}, s: \"6354932Ki\", Format: \"BinarySI\"}, s\"hugepages-2Mi\": {s: \"0\", Format: \"DecimalSI\"}, s\"memory\": {i: {...}, Format: \"BinarySI\"}, ...},\n  \t\tAllocatable: {s\"cpu\": {i: {...}, s: \"2\", Format: \"DecimalSI\"}, s\"ephemeral-storage\": {i: {...}, s: \"6354932Ki\", Format: \"BinarySI\"}, s\"hugepages-2Mi\": {s: \"0\", Format: \"DecimalSI\"}, s\"memory\": {i: {...}, Format: \"BinarySI\"}, ...},\n  \t\tPhase:       \"\",\n  \t\tConditions: []v1.NodeCondition{\n  \t\t\t{\n  \t\t\t\tType:               \"MemoryPressure\",\n  \t\t\t\tStatus:             \"False\",\n- \t\t\t\tLastHeartbeatTime:  v1.Time{Time: s\"2024-05-10 09:42:02 +0000 UTC\"},\n+ \t\t\t\tLastHeartbeatTime:  v1.Time{Time: s\"2024-05-10 09:47:08 +0000 UTC\"},\n  \t\t\t\tLastTransitionTime: {Time: s\"2024-05-08 11:26:11 +0000 UTC\"},\n  \t\t\t\tReason:             \"KubeletHasSufficientMemory\",\n  \t\t\t\tMessage:            \"kubelet has sufficient memory available\",\n  \t\t\t},\n  \t\t\t{\n  \t\t\t\tType:               \"DiskPressure\",\n  \t\t\t\tStatus:             \"False\",\n- \t\t\t\tLastHeartbeatTime:  v1.Time{Time: s\"2024-05-10 09:42:02 +0000 UTC\"},\n+ \t\t\t\tLastHeartbeatTime:  v1.Time{Time: s\"2024-05-10 09:47:08 +0000 UTC\"},\n  \t\t\t\tLastTransitionTime: {Time: s\"2024-05-08 11:26:11 +0000 UTC\"},\n  \t\t\t\tReason:             \"KubeletHasNoDiskPressure\",\n  \t\t\t\tMessage:            \"kubelet has no disk pressure\",\n  \t\t\t},\n  \t\t\t{\n  \t\t\t\tType:               \"PIDPressure\",\n  \t\t\t\tStatus:             \"False\",\n- \t\t\t\tLastHeartbeatTime:  v1.Time{Time: s\"2024-05-10 09:42:02 +0000 UTC\"},\n+ \t\t\t\tLastHeartbeatTime:  v1.Time{Time: s\"2024-05-10 09:47:08 +0000 UTC\"},\n  \t\t\t\tLastTransitionTime: {Time: s\"2024-05-08 11:26:11 +0000 UTC\"},\n  \t\t\t\tReason:             \"KubeletHasSufficientPID\",\n  \t\t\t\tMessage:            \"kubelet has sufficient PID available\",\n  \t\t\t},\n  \t\t\t{\n  \t\t\t\tType:               \"Ready\",\n  \t\t\t\tStatus:             \"True\",\n- \t\t\t\tLastHeartbeatTime:  v1.Time{Time: s\"2024-05-10 09:42:02 +0000 UTC\"},\n+ \t\t\t\tLastHeartbeatTime:  v1.Time{Time: s\"2024-05-10 09:47:08 +0000 UTC\"},\n  \t\t\t\tLastTransitionTime: {Time: s\"2024-05-08 11:26:11 +0000 UTC\"},\n  \t\t\t\tReason:             \"KubeletReady\",\n  \t\t\t\tMessage:            \"kubelet is posting ready status\",\n  \t\t\t},\n  \t\t},\n  \t\tAddresses:       {{Type: \"InternalIP\", Address: \"192.168.39.166\"}, {Type: \"Hostname\", Address: \"minikube\"}},\n  \t\tDaemonEndpoints: {KubeletEndpoint: {Port: 10250}},\n  \t\t... // 5 identical fields\n  \t},\n  }\n"

```